src/ds/logger.h

```diff
@@ -29,6 +31,152 @@ namespace logger
 
   static constexpr size_t ns_per_s = 1'000'000'000;
 
+  class AbstractLogger
+  {
+  protected:
+    std::string log_path;
+    std::ofstream f;
+
+  public:
+    AbstractLogger() = default;
+    AbstractLogger(std::string log_path_) : log_path(log_path_)
+    {
+      f.open(log_path, std::ios_base::app);
+    }
+    virtual ~AbstractLogger() = default;
+
+    std::string get_timestamp(const std::tm& tm, const ::timespec& ts)
+    {
+      // Sample: "2019-07-19 18:53:25.690267"
+      return fmt::format("{:%Y-%m-%dT%H:%M:%S}.{:0<6}Z", tm, ts.tv_nsec / 1000);
+    }
+
+    virtual std::string format(
+      const std::string& file_name,
+      size_t line_number,
+      const std::string& log_level,
+      const std::string& msg,
+      const std::tm& host_tm,
+      const ::timespec& host_ts,
+      const std::optional<::timespec>& enclave_ts = std::nullopt) = 0;
+
+    virtual void write(const std::string& log_line) = 0;
+
+    void dump(const std::string& msg)
+    {
+      f << msg << std::endl;
+      if (!f)
+        throw std::logic_error("Failed to write to file: " + log_path);
+    }
+  };
+
+  class JsonLogger : public AbstractLogger
+  {
+  public:
+    JsonLogger(std::string log_path) : AbstractLogger(log_path) {}
+
+    std::string format(
+      const std::string& file_name,
+      size_t line_number,
+      const std::string& log_level,
+      const std::string& msg,
+      const std::tm& host_tm,
+      const ::timespec& host_ts,
+      const std::optional<::timespec>& enclave_ts = std::nullopt) override
+    {
+      nlohmann::json j;
```

Comment:
If we've removed the other json entries for performance, why do we still have this one? We can use `msg` directly rather than `j["m"].dump()`?

Reply:
because otherwise it gets formatted as a normal string and if it has, for example, new lines in it then it breaks the json format. any other ideas welcome.

Reply:
`\"msg\":\"{}\"` should work? I think all `.dump()` is doing is adding quotes.

Reply:
I tried that it does not work

Reply:
Also you might not only have newline, you might have other invalid characters that you want to be serialized as dump() returns a string containing the serialization of the JSON value